bundles/config/org.eclipse.smarthome.config.xml/src/main/java/org/eclipse/smarthome/config/xml/AbstractXmlBasedProvider.java

```diff
@@ -0,0 +1,231 @@
+        public int hashCode() {
+            final int prime = 31;
+            int result = 1;
+            result = prime * result + ((id == null) ? 0 : id.hashCode());
+            result = prime * result + ((locale == null) ? 0 : locale.hashCode());
+            return result;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) {
+                return true;
+            }
+            if (obj == null) {
+                return false;
+            }
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            LocalizedKey other = (LocalizedKey) obj;
+            if (!Objects.equals(id, other.id)) {
+                return false;
+            }
+            if (!Objects.equals(locale, other.locale)) {
+                return false;
+            }
+            return true;
+        }
+
+    }
+
+    private final Map<Bundle, List<T_OBJECT>> bundleObjectMap = new ConcurrentHashMap<>();
+    private final Map<LocalizedKey, T_OBJECT> localizedObjectCache = new ConcurrentHashMap<>();
+
+    /**
+     * Return the unique identifier of the given object (e.g. UID, URI,...)
+     *
+     * @param object the object to return the identifier for
+     * @return the identifier. Must not be <code>null</code>
+     */
+    protected abstract T_ID getIndentifier(T_OBJECT object);
+
+    /**
+     * Create a translated/localized copy of the given object.
+     *
+     * @param bundle the module to be used for the look-up of the translations
+     * @param object the object to translate
+     * @param locale the target locale
+     * @return a translated copy of the given object or <code>null</code> if translation was not possible.
+     */
+    protected abstract T_OBJECT localize(Bundle bundle, T_OBJECT object, Locale locale);
+
+    /**
+     * Adds an object to the internal list associated with the specified module.
+     * <p>
+     * This method returns silently, if any of the parameters is {@code null}.
+     *
+     * @param bundle the module to which the object is to be added
+     * @param object the object to be added
+     */
+    public final synchronized void add(Bundle bundle, T_OBJECT object) {
+        addAll(bundle, Collections.singletonList(object));
+    }
+
+    /**
+     * Adds a {@link Collection} of objects to the internal list associated with the specified module.
+     * <p>
+     * This method returns silently, if any of the parameters is {@code null}.
+     *
+     * @param bundle the module to which the object is to be added
+     * @param objectList the objects to be added
+     */
+    public final synchronized void addAll(Bundle bundle, Collection<T_OBJECT> objectList) {
+        if (objectList == null || objectList.isEmpty()) {
+            return;
+        }
+        List<T_OBJECT> objects = acquireObjects(bundle);
+        if (objects == null) {
+            return;
+        }
+        for (T_OBJECT object : objectList) {
```

Comment:
You are using an CopyOnWriteArrayList, but also for any other collection, I think it would be "better" to use `objects.addAll(objectList)` (so the collection implementation can perform an optimized algorithm to add multiple objects) and only removeCachedEntries on the for-each-loop.